src/realm/index_string.cpp

```diff
@@ -963,19 +994,32 @@ bool StringIndex::leaf_insert(ObjKey obj_key, key_type key, size_t offset, Strin
     get_child(*m_array, 0, keys);
     REALM_ASSERT(m_array->size() == keys.size() + 1);
 
+    // If we are keeping the complete string in the index
+    // we want to know if this is the last part
+    bool is_at_string_end = offset + 4 >= index_data.size();
+
     size_t ins_pos = keys.lower_bound_int(key);
+    size_t ins_pos_refs = ins_pos + 1; // first entry in refs points to offsets
+
     if (ins_pos == keys.size()) {
         if (noextend)
             return false;
 
         // When key is outside current range, we can just add it
         keys.add(key);
-        int64_t shifted = int64_t((uint64_t(obj_key.value) << 1) + 1); // shift to indicate literal
-        m_array->add(shifted);
+        if (!m_target_column.is_fulltext() || is_at_string_end) {
```

Comment:
It feels to me like we could use a separate class for fulltext search (or possibly a subclass of StringIndex?) There is some  performance penalty (small) to all existing search indexes because they have to check what type of index the operations are for all over the place. Additionally, I expect we will be improving the fulltext features over time, and it would be nice to have that code contained in a specialized place. What do you think?

Reply:
We could create a whole new class, but that would just mean a lot of duplicated code. We could possibly use subclassing, but that would just replace the testing with some virtual method call - which is probably slower.

Reply:
That's true, ok I'm fine with leaving it as is for now.